### PR TITLE
Self in Asset Trait

### LINKED ISSUE #22: Use Self instead of Asset in functions in Asset struct trait
Use Self instead of Asset in functions in Asset struct trait

### PR DESCRIPTION
**Pull request by nvgrw**
Closes #22 

Use Self instead of Asset in functions in Asset struct trait.

This PR fixes the implementation for `Self` types to work in initialisers and (temporarily) allows inits to be mutating as otherwise the `Self` trait won't work. The latter part requires further design to properly implement.

# TODO
- [x] PR description
- [x] Adding `Self` to standard library's Asset struct
- [x] Changing that struct to a struct trait
- [x] Fixing selfs in init
- [x] Fixing mutating init
- [x] Testing

### CODE REVIEW

**Comment by nvgrw** [comment]
Blocked by #30, #56 56

**Comment by nvgrw** [comment]
Now blocked on calling functions in `init`.

We need this to be possible because we want default implementations to be able to set the value of an asset without direct access to stored properties (as traits don't support this). Will investigate further tomorrow.

File: Sources/flintc/main.swift [hunk 1]
```diff
@@ -14,8 +14,9 @@ func main() {
     Flag("dump-ast", flag: "a", description: "Print the abstract syntax tree of the code."),
     Flag("verify", flag: "v", description: "Verify expected diagnostics were produced."),
     Flag("quiet", flag: "q", description: "Supress warnings and only emit fatal errors."),
+    Flag("no-stdlib", flag: "S", description: "Do not load the standard library"),
```
Comment (Aurel300) [inline Sources/flintc/main.swift]: I would get rid of the short flags unless they are 100% obvious (e.g. `-v` is `--version`).

File: Sources/AST/Environment/Environment+Checks.swift [hunk 2]
```diff
@@ -156,14 +156,18 @@ extension Environment {
...
+  // but different modifiers, label names, modifiers, etc
   public func conflictingTraitSignatures(for type: RawTypeIdentifier) -> [String: [FunctionInformation]] {
     guard let typeInfo = types[type] else {
       return [:]
     }
     return typeInfo.traitFunctions.filter { (_, functions) in
       functions.count > 1 && functions.contains(where: {
-        $0.declaration.signature != functions.first!.declaration.signature
+        let compare = $0.declaration.signature
+        let against = functions.first!.declaration.signature
```
[the remark this answers is not in the document]
Comment (nvgrw) [reply Sources/AST/Environment/Environment+Checks.swift]: Doné https://github.com/flintrocks/flint/issues/63

File: Sources/flintc/main.swift (hunk 1 above)
Comment (nvgrw) [reply Sources/flintc/main.swift]: #64

**Pull request merged**